src/API/Run.hs

```diff
@@ -15,39 +15,64 @@ import Parser.Parser
 import Language.Syntax
 import Language.Types
 import Text.Parsec.Pos
-import Text.Parsec (errorPos)
 import Text.Parsec.Error
 
 import Typechecker.Typechecker
+import Typechecker.Monad (Env(types))
 import Runtime.Eval
 import Runtime.Values
 import Runtime.Monad
 
+import Utils.General
+
 -- | Runs BoGL code from raw text with the given commands
 -- utilizes parsePreludeAndGameText to parse the code directly,
 -- without reading it from a file first
 _runCodeWithCommands :: SpielCommand -> IO SpielResponses
 _runCodeWithCommands sc@(SpielCommand _prelude gameFile _ _ filename) =
   (_handleParsed sc $ parsePreludeAndGameText _prelude gameFile filename)
 
-
 -- | Handles result of parsing a prelude and game
 _handleParsed :: SpielCommand -> IO (Either ParseError (Game SourcePos)) -> IO SpielResponses
 _handleParsed (SpielCommand _ gameFile inpt buf _) parsed = do
   pparsed <- parsed
   case pparsed of
     Right game -> do
-      let checked = tc game
-      if success checked
-        then return $ [SpielTypes (rtypes checked), (serverRepl game gameFile inpt (buf, [], 1))]
-        else return $ SpielTypes (rtypes checked) : map (SpielTypeError . snd) (errors checked)
-    Left _err -> do
-      let position = errorPos _err
-          l = sourceLine position
-          c = sourceColumn position
-          in
-        return $ [SpielParseError l c gameFile (show _err)]
+      let progTCRes  = tc game
+          goodResp   = [SpielTypes (rtypes progTCRes), (serverRepl game gameFile inpt (buf, [], 1))]
+          progErrTC  = SpielTypes (rtypes progTCRes) : map (SpielTypeError . snd) (errors progTCRes)
+
+          -- this is bad. We grab the last Val off the buffer, which has not been typechecked
+          -- then we do: Val => String => Expr so that we can type check it
+          -- it would be much better to receive it as a Maybe Expr separate from the buffer
```

Comment:
What we have now works, but is quite sketchy. We can discuss my proposed change for a further change in the near future.

Reply:
Sketchy how so?

Reply:
The comment above is the short version. It would be much easier to explain in full detail over a Discord chat, which we should do soon!